### PR TITLE
Issue #4495 ReservedThreadExecutor optimise

### PR DESCRIPTION
**Pull request by gregw**
Use synchronousQueue for task handoff for #4495 

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Review by sbordet — COMMENTED** [review]
I would hold this until we have results from the 2 clients that are trying these changes.

**Comment by gregw** [comment]
it also would need the alternative implementations removed before merging.

**Comment by gregw** [comment]
@sbordet nudge

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java [hunk 1]
```diff
@@ -264,20 +265,25 @@ public String toString()
...
             if (LOG.isDebugEnabled())
                 LOG.debug("{} offer {}", this, task);
 
-            try (Locker.Lock lock = _locker.lock())
+            try
+            {
+                _task.put(task);
+                return true;
+            }
+            catch (InterruptedException e)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java]: Better catch `Throwable` here.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java [hunk 2]
```diff
@@ -296,37 +302,12 @@ private Runnable reservedWait()
...
-                    }
-                    else
-                    {
-                        Runnable task = _task;
-                        _task = null;
-
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("{} task={}", this, task);
-
+                    Runnable task = _idleTime == 0 ? _task.take() : _task.poll(_idleTime, _idleTimeUnit);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java]: Should be `_idleTime <= 0`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java [hunk 3]
```diff
@@ -296,37 +302,12 @@ private Runnable reservedWait()
...
+                    if (task != null)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java:310]: There was a `LOG.debug()` statement when we had a non-null `task` that I would like to keep.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Comment by gregw** [comment]
@sbordet nudge

**Review by sbordet — APPROVED** [review]

**Pull request merged**